### PR TITLE
MNT: make _setattr_cm more conservative

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

 - special case methods because they are descriptors
 - fail on any other non-instance attribute

This is a simpler version of  #17561

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant

### CODE REVIEW

**Comment by QuLogic** [comment]
CI does not like this at all.

**Comment by tacaswell** [comment]
Indeed, I thought I ran the tests before I pushed.  I must have lost track of what I was doing locally...

**Comment by tacaswell** [comment]
:facepalm: I edited one working tree and installed and tested a different one.

File: lib/matplotlib/tests/test_cbook.py [hunk 1]
```diff
@@ -651,3 +651,68 @@ def test_check_shape(target, test_shape):
...
+
+    def verify_pre_post_state(obj):
+        # When you access a Python method the function is bound
+        # to the object at access time so you get a new instance
+        # of MethodType every time.
+        #
+        # https://docs.python.org/3/howto/descriptor.html#functions-and-methods
+        assert obj.meth is not obj.meth
+        # normal attribute should give you back the same
+        # instance every time
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_cbook.py:688]: Fits on one line.
```suggestion
        # normal attribute should give you back the same instance every time
```

File: lib/matplotlib/cbook/__init__.py [hunk 2]
```diff
@@ -2031,14 +2031,35 @@ def _setattr_cm(obj, **kwargs):
     """
     Temporarily set some attributes; restore original state at context exit.
     """
+    import inspect
     sentinel = object()
-    origs = [(attr, getattr(obj, attr, sentinel)) for attr in kwargs]
+    origs = {}
+    for attr in kwargs:
+        orig = getattr(obj, attr, sentinel)
+        # monkey patching on a new attribute, this is OK
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py:2039]: perhaps you can reorder this to make the common case first, e.g.
```
if (attr in obj.__dict__ or orig is sentinel
        or isinstance(getattr(type(obj), attr), (property, types.FunctionType)):
    origs[attr] = orig
else:
    raise ValueError(...)
```
Checking to types.FunctionType in the type is the same as checking for methods, and (as noted in the other PR, I think) won't give a wrong result if an unrelated bound method is assigned as instance var (`self.foo = otherobj.method`) -- at least I think this point should be fixed.

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py:2039]: That makes sense. I agree the method from another object is a valid concern (sorry I didn't fully register it before).

File: lib/matplotlib/cbook/__init__.py [hunk 3]
```diff
@@ -2032,13 +2032,30 @@ def _setattr_cm(obj, **kwargs):
     Temporarily set some attributes; restore original state at context exit.
     """
     sentinel = object()
-    origs = [(attr, getattr(obj, attr, sentinel)) for attr in kwargs]
+    origs = {}
+    for attr in kwargs:
+        orig = getattr(obj, attr, sentinel)
+
+        if (attr in obj.__dict__ or orig is sentinel):
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py:2039]: no need for parentheses

File: lib/matplotlib/tests/test_cbook.py [hunk 4]
```diff
@@ -651,3 +651,75 @@ def test_check_shape(target, test_shape):
...
+    a = B()
+
+    a.monkey = other.meth
+    verify_pre_post_state(a)
+    with cbook._setattr_cm(
+            a, prop='squirrel',
+            aardvark='moose', meth=lambda: None,
+            override='boo', extra='extra',
+            monkey=lambda: None
+
```
Comment (anntzer) [inline lib/matplotlib/tests/test_cbook.py:709]: blank line a bit weird, and I guess you could either add a trailing comma after None or put the closing parentheses on the same line.

File: lib/matplotlib/tests/test_cbook.py [hunk 5]
```diff
@@ -651,3 +651,75 @@ def test_check_shape(target, test_shape):
...
+        assert a.override == 'boo'
+        assert a.extra == 'extra'
+        assert a.prop == 'squirrel'
+        assert a.monkey != other.meth
+
+    verify_pre_post_state(a)
+
+    with pytest.raises(ValueError):
+        with cbook._setattr_cm(a, cls_level='bob'):
+            ...
```
Comment (anntzer) [inline lib/matplotlib/tests/test_cbook.py:725]: "pass", I guess

**Pull request merged**